proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,111 @@
+
+## Introduction
+
+It is increasingly important to build fault tolerant micro services. Fault tolerance is about leveraging different strategies to guide the execution and result of some logic. Retry policies, bulkheads, and circuit breakers are popular concepts in this area. They dictate whether and when executions should take place, and fallbacks offer an alternative result when an execution does not complete successfully. 
+
+As mentioned above, the Fault Tolerance proposal is to focus the aspects: TimeOut, RetryPolicy, Fallback, bulkhead and circuit breaker.
+
+* TimeOut: Define a duration for timeout
+* RetryPolicy: Define a criteria on when to retry 
+* Fallback: provide an alternative solution for a failed execution.
+* Bulkhead: isolate failures in part of the system while the rest part of the system can still function.
+* CircuitBreaker: offer a way of fail fast by automatically failing execution to prevent the system overloading and indefinite wait or timeout by the clients.
+
+The main design is to separate execution logic from execution. The execution can be configured with fault tolerance policies, such as RetryPolicy, fallback, Bulkhead and CircuitBreaker. 
+
+[Hystrix](https://github.com/Netflix/Hystrix) and [Failsafe](https://github.com/jhalterman/failsafe) are two popular libraries for handling failures. This proposal is to define a standard API and approach for applications to follow in order to achieve the fault tolerance.
+
+The requirements are as follows:
+
+* Loose coupling: Execution logic should not know anything about the execution status or fault tolerance. 
+* Failure handling strategy should be configured when the execution takes place.
+* Support for synchronous and asynchronous execution
+* Integration with 3rd party asynchronous APIs. This is necessary to handle executions that are completed at some time in the future, where retries will need to be explicitly scheduled from within the asynchronous execution. This is common when working with various 3rd party asynchronous tools such as Netty, RxJava, Vert.x, etc.
+* Support immutable failure handling policy configuration
+* Some Failure policy configurations, e.g. CircuitBreaker, RetryPolicy, can be used stand alone. For example, it has been very useful for circuit breakers to be standalone constructs which can be plugged into and intentionally shared across multiple executions. Likewise for retry policies. Additionally, an Execution construct can be offered that allows retry policies to be applied to some logic in a standalone, manually controlled way.
+
+Advanced requirements:
+
+* Event Hooks: Since this approach to fault tolerance involves handing execution over to some foreign code, it's very useful to be able to learn when executions are taking place and under what circunstances (onRetry, onFailedAttempt, onFailure, etc).
```

Comment:
@jhalterman 
In the case of *supervision* the handler for a failure returns the action to be performed in response. (as with your last example)
(Noting also that an Actor's *Receive* method is actually a method which *produces* a handler for messages rather than directly handling messages itself)